compiler/src/main/scala/org/scalajs/core/compiler/PrepWasPublicBeforeTyper.scala

```diff
@@ -0,0 +1,72 @@
+/* Scala.js compiler
+ * Copyright 2013 LAMP/EPFL
+ * @author Nicolas Stucki
+ */
+
+package org.scalajs.core.compiler
+
+import scala.tools.nsc
+import nsc._
+
+/**
+ * @author Nicolas Stucki
+ */
+abstract class PrepWasPublicBeforeTyper extends plugins.PluginComponent with transform.Transform
+    with Compat210Component {
+
+  import global._
+
+  val phaseName = "prepwaspublicbeforetyper"
+
+  override protected def newTransformer(unit: CompilationUnit): Transformer =
+    new PreTyperTransformer(unit)
+
+  class PreTyperTransformer(unit: CompilationUnit) extends Transformer {
+    override def transform(tree: Tree): Tree = tree match {
+      case tree: ClassDef if needsAnnotations(tree) =>
+        val newBody = tree.impl.body.map {
+          case vdef: ValDef if needsAnnotations(vdef) =>
+            treeCopy.ValDef(vdef, withWasPublic(vdef.mods), vdef.name,
+                vdef.tpt, vdef.rhs)
+
+          case ddef: DefDef if needsAnnotations(ddef) =>
+            treeCopy.DefDef(ddef, withWasPublic(ddef.mods), ddef.name,
+                ddef.tparams, ddef.vparamss, ddef.tpt, ddef.rhs)
+
+          case member => member
+        }
+        val newImpl =
+          treeCopy.Template(tree.impl, tree.impl.parents, tree.impl.self, newBody)
+        val newTree =
+          treeCopy.ClassDef(tree, tree.mods, tree.name, tree.tparams, newImpl)
+        super.transform(newTree)
+
+      case _ => super.transform(tree)
+    }
+  }
+
+  private def needsAnnotations(classDef: ClassDef): Boolean = {
+    classDef.name.toString == "$anon" &&
+    classDef.impl.body.exists {
+      case vdef: ValDef => needsAnnotations(vdef)
+      case ddef: DefDef => needsAnnotations(ddef)
+      case _ => false
+    }
+  }
+
+  private def needsAnnotations(vdef: ValDef): Boolean =
+    vdef.mods.isPublic
+
+  private def needsAnnotations(ddef: DefDef): Boolean =
+    ddef.mods.isPublic && ddef.name != nme.CONSTRUCTOR
+
+  private def withWasPublic(mods: Modifiers): Modifiers =
+    mods.withAnnotations(anonymousClassMethodWasPublicAnnotations)
+
+  private lazy val anonymousClassMethodWasPublicAnnotations: List[Tree] = {
+    val runtimePackage =
+      Select(Select(Ident(newTermName("scala")), newTermName("scalajs")), newTermName("runtime"))
+    val cls = Select(runtimePackage, newTypeName("WasPublicBeforeTyper"))
+    List(Apply(Select(New(cls), nme.CONSTRUCTOR), Nil))
```

Comment:
I think it's risky to reuse Trees. I suggest making this a `def` to create new Trees every time. (the `newTermName`s could be cached, though)